agent/acs/handler/refresh_credentials_responder_test.go

```diff
@@ -0,0 +1,275 @@
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package handler
+
+import (
+	"fmt"
+	"testing"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	mock_engine "github.com/aws/amazon-ecs-agent/agent/engine/mocks"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	acssession "github.com/aws/amazon-ecs-agent/ecs-agent/acs/session"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/session/testconst"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/credentials"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/golang/mock/gomock"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	expiration   = "soon"
+	roleArn      = "taskrole1"
+	accessKey    = "akid"
+	secretKey    = "secret"
+	sessionToken = "token"
+	roleType     = "TaskExecution"
+)
+
+var expectedCredentialsAck = &ecsacs.IAMRoleCredentialsAckRequest{
+	Expiration:    aws.String(expiration),
+	MessageId:     aws.String(testconst.MessageID),
+	CredentialsId: aws.String(testconst.CredentialsID),
+}
+
+var expectedCredentials = credentials.TaskIAMRoleCredentials{
+	ARN: testconst.TaskARN,
+	IAMRoleCredentials: credentials.IAMRoleCredentials{
+		RoleArn:         roleArn,
+		AccessKeyID:     accessKey,
+		SecretAccessKey: secretKey,
+		SessionToken:    sessionToken,
+		Expiration:      expiration,
+		CredentialsID:   testconst.CredentialsID,
+		RoleType:        roleType,
+	},
+}
+
+var testRefreshCredentialsMessage = &ecsacs.IAMRoleCredentialsMessage{
+	MessageId: aws.String(testconst.MessageID),
+	TaskArn:   aws.String(testconst.TaskARN),
+	RoleType:  aws.String(roleType),
+	RoleCredentials: &ecsacs.IAMRoleCredentials{
+		RoleArn:         aws.String(roleArn),
+		Expiration:      aws.String(expiration),
+		AccessKeyId:     aws.String(accessKey),
+		SecretAccessKey: aws.String(secretKey),
+		SessionToken:    aws.String(sessionToken),
+		CredentialsId:   aws.String(testconst.CredentialsID),
+	},
+}
+
+// TestInvalidCredentialsMessageNotAcked tests that invalid credential message
+// is not ACKed.
+func TestInvalidCredentialsMessageNotAcked(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ackSent := false
+	testResponseSender := func(response interface{}) error {
+		ackSent = true
+		return nil
+	}
+	testRefreshCredentialsResponder := acssession.NewRefreshCredentialsResponder(credentials.NewManager(),
+		&credentialsMetadataSetter{
+			taskEngine: nil,
+		},
+		metrics.NewNopEntryFactory(),
+		testResponseSender)
+
+	handleCredentialsMessage := testRefreshCredentialsResponder.HandlerFunc().(func(*ecsacs.IAMRoleCredentialsMessage))
```

Comment:
Why is the casting needed?

Reply:
Here we do not have access to the private refresh credentials responder method `handleCredentialsMessage`. The `acssession.NewRefreshCredentialsResponder` function return type is `wsclient.RequestResponder`, so we have access to the returned value's `HandlerFunc()` method, who's return type is `wsclient.RequestHandler`. 

We do a type assertion here to assert that we expect the `wsclient.RequestHandler` returned by the responder's `HandlerFunc()` method to be a function that only accepts `IAMRoleCredentialsMessage`s. If this type assertion fails, something has gone wrong and we will be notified.

Reply:
Thanks. I didn't realize that `refreshCredentialsResponder` is private and its constructor returns a `wsclient.RequestResponder`. It's a bit odd to make the implementation private but then also use casting here so that we can call its handler function. I noticed that the type of the handler function is `type RequestHandler interface{}` which has no type information... 

Do we use such casting in prod usage of responders too?

Reply:
Discussed offline with @danehlim. We found out that in Prod we use reflection to perform type checks on handler functions provided by ACS responders. So, I am okay with us using type assertions in this test. 

In general I think reflection is a code smell and we should be able to use generics to make the design statically typed, however, that is out of scope of what we are trying to accomplish with this PR. 

https://github.com/aws/amazon-ecs-agent/blob/6ba7f1ea301df8179d8bdf96368332a92a442375/ecs-agent/wsclient/client.go#L373C1-L384